### DISCUSSION TITLE
Mirrors are not generated for Case Classes with multiple parameter lists

### DISCUSSION THREAD

**Discussion by Lasering** [scala/scala3#17778] [Feature Requests]
## Compiler version

3.1.1

## Minimized code
```Scala
import scala.deriving.Mirror

case class Foo(a: Int, b: Int)
case class Bar1(a: Int)(b: Int)
case class Bar2(a: Int)(val b: Int)

summon[Mirror.Of[Foo]]
summon[Mirror.Of[Bar1]]
summon[Mirror.Of[Bar2]]
```

[Scastie Link](https://scastie.scala-lang.org/hNQajG94RuOmAC6IG7NJ2w)

## Output
From scastie
```scala
no implicit argument of type deriving.Mirror.Of[Playground.Bar1] was found for parameter x of method summon in object Predef
no implicit argument of type deriving.Mirror.Of[Playground.Bar2] was found for parameter x of method summon in object Predef
```

## Expectation
A Mirror should be generated for Bar1 and Bar2

**Comment by bishabosha** [comment]
this is as expected the compiler. There is an explicit check for secondary argument lists in the constructor. The main problem is how do we implement a correct `fromProduct` method in the mirror for these cases. Also what would you expect for the `MirroredElemTypes` to look?

**Comment by Lasering** [comment]
Ideally a tuple of tuples mapping to the parameter lists.

**Comment by joroKr21** [comment]
But secondary parameter lists don't participate in `hashCode`, `equals`, `unapply` - I don't think it's a good idea in general

**Comment by joroKr21** [comment]
Shapeless also has issues with this. A related problem is what happens with implicits like `case class Foo[A: Order](value: A, label: String)` - Shapeless 2 would capture the implicit `Order[x]` at the place where `Generic[Foo[x]]` is summoned. Which works but can be a gotcha.

**Comment by Lasering** [comment]
> But secondary parameter lists don't participate in `hashCode`, `equals`, `unapply` - I don't think it's a good idea in general

For example when deriving a Decoder for a class with multiple parameter lists I want to be able to create an instance of that class. Without knowing the types of the remaining parameter lists there is no way to do that.

It is known that parameters (`val`s or not) of secondary parameter lists are not used for the automatically generated `hashCode`, `equals`, `unapply`. Can you explain how that is affected/impacted by the inclusion of the secondary parameter lists on the Mirror?

**Comment by joroKr21** [comment]
Consider that `Mirror.Product` is based on two functions - `fromProduct` and `toProduct`. Well `toProduct` for a case class is just identity because case classes are already products. However the fields that are considered part of this product are only the primary fields. Now if you include the secondary fields to `fromProduct` you get an inconsistency - it would not work correctly (i.e. `fromProduct` after `toProduct` no longer gives the original case class back.

And then there is the interaction with implicits that I already mentioned.

**Comment by Lasering** [comment]
To make things consistent `toProduct` would have to also produce a tuple of tuples. But I'm still missing the relation to `hashCode`, `equal`, `unapply` are those implemented using the result of `toProduct`?

Is there other avenues that could be pursued? Maybe a specialized `Mirror` or `toProductAll`/`fromProductAll`

**Comment by joroKr21** [comment]
All this stuff should be consistent (also with Mirror):
```scala
scala> case class Foo(x: Int)(val y: String)
// defined case class Foo

scala> val foo = Foo(42)("bar")
val foo: Foo = Foo(42)

scala> foo.hashCode
val res1: Int = 759270620

scala> import scala.util.hashing.MurmurHash3

scala> MurmurHash3.productHash(foo)
val res2: Int = 759270620

scala> foo.productArity
val res3: Int = 1

scala> foo.productIterator.toList
val res4: List[Any] = List(42)
```

**Comment by odersky** [comment]
I think all this goes too far. We have already exhausted the complexity and code size budget for mirrors and related stuff. I'll move to feature request, but don't think this will be adopted in the end.
